inspirehep/modules/orcid/api.py

```diff
@@ -97,39 +105,58 @@ def push_record_with_orcid(recid, orcid, oauth_token, put_code=None, old_hash=No
                 token=oauth_token,
                 request_type='work',
                 data=orcid_xml,
-                put_code=put_code,
+                put_code=putcode,
                 content_type='application/orcid+xml',
             )
+
+    # It's a new record: POST.
     else:
         LOGGER.info(
             "No put-code found, pushing new record #%s to ORCID %s.", recid, orcid,
         )
-        with log_time_context('Pushing new record', LOGGER), \
-                distributed_lock(lock_name, blocking=True):
-            put_code = orcid_api.add_record(
-                orcid_id=orcid,
-                token=oauth_token,
-                request_type='work',
-                data=orcid_xml,
-                content_type='application/orcid+xml',
-            )
-
-    LOGGER.info("Push of %s onto %s completed with put-code %s.", recid, orcid, put_code)
 
-    return put_code, new_hash
-
-
-def calculate_hash_for_record(record):
-    """Generate hash for an ORCID-serialised HEP record
+        try:
+            with log_time_context('Pushing new record', LOGGER), \
+                    distributed_lock(lock_name, blocking=True):
+                putcode = orcid_api.add_record(
+                    orcid_id=orcid,
+                    token=oauth_token,
+                    request_type='work',
+                    data=orcid_xml,
+                    content_type='application/orcid+xml',
+                )
+        except Exception as exc:
+            if DuplicatedExternalIdentifiersError.match(exc):
+                recache_author_putcodes(orcid, oauth_token)
+                putcode = orcid_cache.read_work_putcode(recid)
+                if not putcode:
+                    raise PutcodeNotFoundInCacheException(
+                        'Putcode not found in cache for recid {} after having'
+                        ' recached all putcodes for the orcid {}'.format(recid, orcid))
+                return push_record_with_orcid(recid, orcid, oauth_token)
+            else:
+                raise exc
```

Comment:
I think it's better to use raise without `exc`. In Python 2 `raise exc` will lose the original traceback: https://stackoverflow.com/a/4825279. Unless this is what you want

Reply:
Good point, what about using the `raise_from` from six? https://pythonhosted.org/six/#six.raise_from